glossary.yml

```diff
@@ -461,11 +473,14 @@
   en:
     term: "assertion"
     def: >
-      A [Boolean](#boolean) expression that must be true at a certain point in a
+      A [Boolean](#boolean) expression that must be [true](#true) at a certain point in a
       program. Assertions may be built into the language (e.g., [Python](#python)'s
-      `assert` statement) or provided as functions (e.g., R's `stopifnot`). They are
+      `assert` statement) or provided as functions (e.g., [R's](#r_language) `stopifnot`). They are
       often used in testing, but are also put in [production code](#production_code)
-      to check that it is behaving correctly.
+      to check that it is behaving correctly. Assertions should not be used to perform
```

Comment:
I think R may be an exception here. I don't know, I don't use it. And, to be clear, I am saying assertions should not be used for **data validation** in production code. Using them for testing is fine, but in production code they should only identify things that truly break code. *There are other ways to validate data.*

I feel very strongly about this. [Here is a good summary of why; see Caveat #1: Don't Use Asserts for Data Validation](https://dbader.org/blog/python-assert-tutorial).

Python, and several other languages have an option to optimise assertions away. (I consulted an actual computer science professor on that point when I was correcting this entry, just to make sure this wasn't the minority case.) In Python, assertions are on by default, but turned off under any optimisation condition, but in Java, this is the other way around—assertions must be explicitly enabled (source, Deitel & Deitel 10th edition, Java: How to Program). But, basically, if you rely on data validation to verify the output of a program, and that data validation can be disabled, that seems like really bad practice, particularly if anyone else uses the code and that point isn't very clearly communicated.

There is also a Python lesson that also teaches assertions, and advocates using them for data validation, but that doesn't make it right. I have submitted an [issue](https://github.com/swcarpentry/python-novice-inflammation/issues/874) about it, talked to the maintainers in the issue comments, and am working on a pull request to fix it.

I would be willing to compromise with something citing R as an exception, or saying that 'In many languages, assertions should not be used...,' but I think the point is important to mention.

Reply:
After reading the post, I think R is indeed an exception because R does not have the concept of an assert statement in the same sense as other languages since there is no way to ignore these commands with a flag to the interpreter.

R is a bit of an odd language probably because it was built by statisticians. I think it's safe to make a note citing R as an exception to the rule or even saying that R does not have an assert statement.

Reply:
Yes, I wanted to say 'R is generally a poor choice as a *model* of a programming language, but some R people don't like it when I say things like that.

(i'm glad you clarified there is not an assert statement in R, because I was looking in my books for any reference, and not finding anything.)